src/sentry/workflow_engine/models/detector.py

```diff
@@ -129,6 +127,74 @@ def evaluate(self, data_packet: DataPacket[T]) -> list[DetectorEvaluationResult]
 
 
 class StatefulDetectorHandler(DetectorHandler[T], abc.ABC):
+    @property
+    @abc.abstractmethod
+    def counter_names(self) -> list[str]:
+        """
+        The names of counters that this detector is going to keep track of.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_dedupe_value(self, data_packet: DataPacket[T]) -> int:
+        """
+        Extracts the deduplication value from a passed data packet.
+        TODO: This might belong on the `DataPacket` instead.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_group_key_values(self, data_packet: DataPacket[T]) -> dict[str, int]:
+        """
+        Extracts the values for all the group keys that exist in the given data packet,
+        and returns then as a dict keyed by group_key.
+        """
+        pass
+
+    def get_state_data(self, group_keys: list[str | None]) -> dict[str | None, DetectorStateData]:
+        """
+        Fetches state data associated with this detector for the associated `group_keys`.
+        Returns a dict keyed by each group_key with the fetched `DetectorStateData`.
+        If data isn't currently stored, falls back to default values.
+        """
+        group_key_detectors = self.bulk_get_detector_state(group_keys)
+        dedupe_keys = [self.build_dedupe_value_key(gk) for gk in group_keys]
+        pipeline = get_redis_client().pipeline()
+        for dk in dedupe_keys:
+            pipeline.get(dk)
+        group_key_dedupe_values = {
+            gk: int(dv) if dv else 0 for gk, dv in zip(group_keys, pipeline.execute())
+        }
+        pipeline.reset()
+        counter_updates = {}
+        if self.counter_names:
+            counter_keys = [
+                self.build_counter_value_key(gk, name)
+                for gk in group_keys
+                for name in self.counter_names
+            ]
+            for ck in counter_keys:
+                pipeline.get(ck)
+            vals = [int(val) if val is not None else val for val in pipeline.execute()]
+            counter_updates = {
+                gk: dict(zip(self.counter_names, values))
+                for gk, values in zip(group_keys, chunked(vals, len(self.counter_names)))
+            }
+
+        results = {}
+        for gk in group_keys:
+            detector_state = group_key_detectors.get(gk)
+            results[gk] = DetectorStateData(
+                group_key=gk,
+                active=detector_state.active if detector_state else False,
+                status=DetectorStatus(detector_state.state)
+                if detector_state
+                else DetectorStatus.OK,
```

Comment:
nit: might be easier to read this if we created a `status` variable and passed that in here.